pytorch_lightning/callbacks/early_stopping.py

```diff
@@ -131,11 +131,11 @@ def __init__(
     def state_key(self) -> str:
         return self._generate_state_key(monitor=self.monitor, mode=self.mode)
 
-    def on_pretrain_routine_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
+    def on_init_end(self, trainer: "pl.Trainer") -> None:
         if self._check_on_train_epoch_end is None:
-            # if the user runs validation multiple times per training epoch, we try to check after
-            # validation instead of on train epoch end
-            self._check_on_train_epoch_end = trainer.val_check_interval == 1.0
+            # if the user runs validation multiple times per training epoch or multiple training epochs without
+            # validation, then we run after validation instead of on train epoch end
+            self._check_on_train_epoch_end = trainer.val_check_interval == 1.0 and trainer.check_val_every_n_epoch == 1
```

Comment:
Context:
- Users can log in `training_step` or in `validation_step`.
- We don't which one logs the `monitor` we need.

Now, if we look at `check_val_every_n_epochs>1`:

- Case 1. `training_step` logs the monitor. We want `check_on_train_epoch_end=True`. This is what's in master
- Case 2. `validation_step` logs the monitor. We want `check_on_train_epoch_end=False`. This is what this PR implements.

As you already notice, this would be wrong for case 1 but correct for case 2.

The question then becomes:
Which case do we support by default? I lean towards case 2 as I feel it's more common to early-stop on a validation monitor.

Whichever we chose, the other one will see the error reported in the linked issue until they set the appropriate `check_on_train_epoch_end` value.

Reply:
Thanks, I believe this is acceptable. Approved.